### ISSUE TITLE
Git operations in pure Go

### ISSUE THREAD

**Issue by Thibauth** [gogits/gogs#2592]
If I understand correctly, git operations are actually performed through the gogits/git-module package and shell calls to the git command. Is the long term plan to use gogits/git and native access to the git repository? If so, what is preventing from transitioning to gogits/git? What is the expected performance gain if any?

**Comment by Unknwon** [comment]
Good question, but there is simply no pure go Git implementation.

**Comment by Thibauth** [comment]
Isn't this what gogits/go claim to be? What are the main features missing from it? An alternative solution would be to use git2go (a wrapper around libgit2). But then it is not clear how the overhead of calling C code would compare to a pure implementation.

**Comment by Unknwon** [comment]
gogits/go is a good start but it is not a **fully featured pure go Git implementation**, git2go uses cgo and it is not acceptable right now.

**Comment by nd2s** [comment]
Why would you want to use a (very likely bad) clone of git instead of the original? Makes no sense and just causes trouble and work. 

> Small is beautiful. Small things have tremendous advantages over their larger counterparts. Among these is the ability to combine with other small things in unique and useful ways.
>
> Make each program do one thing well. By focusing on a single task, a program can eliminate much extraneous code that often results in excess overhead, unnecessary complexity, and a lack of flexibility.

http://c2.com/cgi/wiki?UnixDesignPhilosophy

**Comment by Thibauth** [comment]
@n2s, the purpose would be to improve performance and reduce the memory footprint of Gogs, even if this would have to be quantified precisely before deciding to transition. As much as I like the Unix design philosophy, imagining the Gogs process forking and creating a child process on every incoming request violates my taste for simplicity.

Furthermore, designing the pure git library as a separate project would not go against the Unix design philosophy by allowing this separate library to focus only on being good at accessing git repositories :) This is the path that many interfaces for git follow (many use libgit2 or wrappers around it). At least for read access, re-implementing git functionalities is not too hard, and the original git could still be used for the less common write tasks that Gogs requires.

**Comment by emansom** [comment]
When browsing the web I came across this: https://github.com/src-d/go-git

Quote from the README.md file:
"A low level and highly extensible git client library for reading repositories from git servers. It is written in Go from scratch, without any C dependencies."

Might it be an option for Gogs?

**Comment by vtolstov** [comment]
I thnik best of all - create git interface and modify current operations to use this interface.
After that implement other git providers not very hard. @Unknwon what you say?

**Comment by Thibauth** [comment]
@vtolstov, if I understood what you said correctly, I think this is the approach that [git-native-module](https://github.com/mechmind/git-native-module) is following: trying to stay compatible with the interface that gogs requires while re-implementing the internals in pure Go.

**Comment by emansom** [comment]
Please no pure go. It will cause gogs to have issues that will be long solved in libgit2. There's a reason Github and Bitbucket use it in production.

**Comment by Thibauth** [comment]
I see, I think libgit2 (and probably its go wrapper git2go?) is a good alternative and yes, I am also all for not re-inventing the wheel if possible. As a side question, I am curious about how the overhead of calling C code compares to a pure go implementation though.

**Comment by emansom** [comment]
The overhead will be considerably less than the current way of operation; spawning a process.

**Comment by Unknwon** [comment]
I agree with @vtolstov 's idea.

@emansom libgit2 is written in C and it is a plain pain to compile on Windows. Will consider this option but at first we need to make things work, then better.

**Comment by emansom** [comment]
@Unknwon You're sacrificing quality over potential userbase growth now. I don't think that's a wise choice.

**Comment by Unknwon** [comment]
@emansom CGO could not be cross-compiled in Go, that is the real sacrifice quality of potential userbase growth. You would have to compile you own in many platforms.

**Comment by emansom** [comment]
@Unknwon Seems implemented; https://github.com/golang/go/issues/4714. There are also https://github.com/karalabe/xgo and https://github.com/inconshreveable/gonative which might offer a solution.

**Comment by Unknwon** [comment]
@emansom thanks for the pointers, but still, I believe the quality of libgit2 as developed by GitHub, but not in git2go.

Currently, I know where to locate and fix the problem in the current implementation, but switch to git2to, no, no to say they would accept pull request efficiently, that is a road blocker and many other uncertainties. So at least for a long time, current implementation will still exist as the default option.

@vtolstov brings an interesting point that Gogs should abstract the Git operation level, that may save a lot work and allow user to choose between default, git2go, or even others.

**Comment by Thibauth** [comment]
> @vtolstov brings an interesting point that Gogs should abstract the Git operation level, that may save a lot work and allow user to choose between default, git2go, or even others.

I completely agree, I think a good starting point would be to create a list of all the functions from git-module that are called by gogs. Stabilize this as a _git backend_ interface and see if such an interface could easily be implemented using already existing libraries like git2go or pure go implementations. 

I am still new to Go but it seems that the interface/type system of Go is really well-suited for this kind of modular programming.

**Comment by Unknwon** [comment]
@Thibauth yes, you could change the title of this issue to `Abstract the Git operation level` or something similar, but not likely would be able to implemented in the near feature as we still have many tasks ahead.

**Comment by lavvy** [comment]
Please has this been implemented?
